src/java.base/share/classes/java/net/URL.java

```diff
@@ -748,23 +803,70 @@ public URL(URL context, String spec, URLStreamHandler handler)
     }
 
     /**
-     * Creates a URL from a URI, as if by invoking {@code uri.toURL()}.
+     * Creates a URL from a URI, as if by invoking {@code uri.toURL()}, but
+     * associating it with the given {@code URLStreamHandler}, if allowed.
+     *
+     * @apiNote
+     * Applications should consider performing additional integrity
+     * checks before constructing a {@code URL} and opening a connection.
+     * See the <a href=#integrity>API note</a> in the class level API
+     * documentation.
+     *
+     * @implSpec The implementation of this method includes calling the {@link
+     * URLStreamHandler#parseURL(URL, String, int, int) parseURL} method on the
+     * selected handler.
+     *
+     * @param uri the {@code URI} from which the returned {@code URL} should
+     *           be built
+     * @param streamHandler a custom protocol stream handler for
+     *                      the returned {@code URL}. Can be {@code null},
+     *                      in which case the default stream handler for
+     *                      the protocol if any, will be used.
+     *
+     * @return a new {@code URL} instance created from the given {@code URI}
+     *   and associated with the given {@code URLStreamHandler}, if any
+     *
+     * @throws NullPointerException if {@code uri} is {@code null}
+     *
+     * @throws IllegalArgumentException if no protocol is specified
+     *         (the {@linkplain URI#getScheme() uri scheme} is {@code null}), or
+     *         if the {@code URLStreamHandler} is not {@code null} and can not be
+     *         set for the given protocol
+     *
+     * @throws  MalformedURLException if an unknown protocol is found,
+     *          or the given URI fails to comply with the specific
+     *          syntax of the associated protocol, or the
+     *          underlying stream handler's {@linkplain
+     *          URLStreamHandler#parseURL(URL, String, int, int)
+     *          parseURL method} throws {@code IllegalArgumentException}
+     *
+     * @throws SecurityException
+     *        if a security manager exists and its
+     *        {@code checkPermission} method doesn't allow
+     *        specifying a stream handler
      *
      * @see java.net.URI#toURL()
+     *
+     * @since 20
      */
-    static URL fromURI(URI uri) throws MalformedURLException {
+    public static URL fromURI(URI uri, URLStreamHandler streamHandler)
```

Comment:
What do you think to have this method in URI instead: URI.toURL(URLStreamHandler), as there is an URI.toURL() method already?

Reply:
`URLStreamHandler` really belongs to `java.net.URL`, and is an implementation detail of the infrastructure/SPI that makes it possible to eventually call `URL::openConnection`. I do not think it would be appropriate to have such a method in `java.net.URI`. Dealing with `URLStreamHandler` is not something that a regular developer will need to do, and it is better if it isn't too prominent.

Reply:
I see your point.  It may be more appropriate if URI.toURL was designed as URL.fromURL.

I was wondering to have application developers a consistent way to get an URL instance.  Now there are two methods in different classes URI.toURL and URL.fromURI.  It might be easier to use the old URI.toURL form.

Never mind, it is just my personal preference.  It is fine to me to have a new URL.fromURI method.

Reply:
One thing we might do is change the name of the method into `URL.of(URI, StreamHandler)`. It's named 'fromURI' only because there was a pre-existing package protected 'fromURI' method, but since we're making it public now, it could be fair game to change it's name. Possibly adding an overload `URL::of(URI)` method could be considered, but then there really would be two paths to do the same thing - so I'd rather not add such an overload - unless I get some more feedback on that from the CSR/review.

Reply:
I think `URL.of(URI, URLStreamHandler)` is fine.

As for `URL.of(URI)`, whose implementation I suspect will just do `uri.toURL()` on the passed `URI`, I don't think we need it. It might add to unnecesary confusion on whether/when to use `URL.of(URI)` and when to use `URI.toURL()`

In the case of `URL.of(URI, URLStreamHandler)` it's pretty clear that you use it (only) when you have a specific `URLStreamHandler` to use for the constructed `URL`.